// extrade/extrade.js
var crypto = require('crypto');
var querystring = require('querystring');
var request = require('request');

/**
 * exTrade
 *
 * @param {string} key    API Key
 * @param {string} secret API Secret
 */
var exTrade = function(key, secret) {
    this.key = key;
    this.secret = secret;
    this.nonce = Math.floor((new Date()).getTime() / 1000);
    this.userAgent = 'Mozilla/4.0 (compatible; exTrade API Node.js client; NodeJS/' + process.version + ')';

    this.privateApiUrl = 'https://1ex.trade/api/';
    this.publicApiUrl = 'https://1ex.trade/api/';

    this.publicMethods = [
        'historical-prices', 'stats', 'order-book', 'transactions'
    ];

    this.privateMethods = [
        'balances-and-info', 'open-orders', 'user-transactions', 'crypto-deposit-address/get', 'crypto-deposit-address/new', 'deposits/get', 'withdrawals/get',
        'orders/new', 'orders/edit', 'orders/cancel', 'orders/status', 'withdrawals/new'
    ];
}

/**
 * API call
 *
 * @param  {string}   method   API Method
 * @param  {object}   params   Method parameters
 * @param  {Function} callback (err, data)
 */
exTrade.prototype.api = function(method, params, callback) {
    params = params || {};
    params.nonce = ++this.nonce;
    params.api_key = this.key;
    if (this.publicMethods.indexOf(method) >= 0) {
        this.getRequest(params, method, callback);
    } else if (this.privateMethods.indexOf(method) >= 0) {
        this.postRequest(params, method, callback);
    } else {
        callback(new Error('Unknown method: ' + method), null);
    }
}

/**
 * Creates a Base64 signature of the request
 *
 * @param  {string} str Text string
 * @return {string}     Base64 encoded signature
 */
exTrade.prototype.getSignatureFromString = function(str) {
    var hmac = new crypto.createHmac('sha256', this.secret);
    hmac.update(str);
    return hmac.digest('hex');
}

/**
 * HTTP GET request
 *
 * @param  {Object}   params   GET parameters
 * @param  {Function} callback (err, data)
 */
exTrade.prototype.getRequest = function(params, method, callback) {
    var options = {
        url: this.publicApiUrl+method+'?'+querystring.stringify(params, '&', '='),
        method: 'GET',
        headers: {
            'User-Agent': this.userAgent,
        }
    }
    request.get(options, function (err, response, body) {
    	var post = false
        this.parseResponse(err, response, body, post, callback);
    }.bind(this));
}

/**
 * HTTP POST request
 *
 * @param  {Object}   params   POST parameters
 * @param  {Function} callback (err, data)
 */
exTrade.prototype.postRequest = function(params, method, callback) {
	const test = this.getSignatureFromString(querystring.stringify(params));
	console.log(JSON.stringify(params)+'\n');
	params.signature = test;
	console.log(test+'\n');
	console.log(JSON.stringify(params)+'\n');
    var options = {
        url: this.privateApiUrl+method,
        method: 'POST',
        form: params,
        headers: {
            'apisign': test,
            'Key': this.key,
            'User-Agent': this.userAgent,
        }
    };

    request.post(options, function (err, response, body) {
    	var post = true;
        this.parseResponse(err, response, body, post, callback);
    }.bind(this));
}

/**
 * Parse server response
 *
 * @param  {[type]}   err      [description]
 * @param  {[type]}   response [description]
 * @param  {[type]}   body     [description]
 * @param  {Function} callback [description]
 * @return {[type]}            [description]
 */
exTrade.prototype.parseResponse = function(err, response, body, post, callback) {
	if (post){
	    //console.log("Err="+err);
	    //console.log("Response=", response);
	    console.log("Body=", body);
	}
    if (err) {
        callback(new Error('Error in server response: ' + JSON.stringify(err)), null);
    } else {
        if (typeof callback === 'function') {
            var data = JSON.parse(response.body);
            var err = null;
            try {
                data = JSON.parse(response.body);
            } catch(e) {
                err = new Error('Error parsing JSON: ' + response.body);
            }
            if (err) {
                callback(err, data);
            } else {
                if (data) {
                    callback(err, data);
                } else {
                    if (data && data.error) {
                        err = new Error(data.error);
                    } else {
                        err = new Error('Unknown error');
                    }
                    callback(err, null);
                }
            }
        }
    }
}

module.exports = exTrade;
